### PR TITLE
WIP: Support for different scenario

### PR DESCRIPTION
**Pull request by arthurdouillard**
I'm opening this PR so I see more easily the changes, and that we can discuss about them until it's merged. :)

### CODE REVIEW

File: clloader/datasets/continuumDataset.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+import torch
+import numpy as np
+from clloader.viz import plot
+from PIL import Image
+from torchvision import transforms
+from torch.utils.data import Dataset as TorchDataset
+
+
+class ContinuumDataset(TorchDataset):
```
Comment (arthurdouillard) [inline clloader/datasets/continuumDataset.py:11]: It's not really a "continuum dataset" but more a "task dataset" here, isn't it?

File: clloader/loader.py [hunk 2]
```diff
@@ -121,22 +54,33 @@ def __init__(
...
-        nb_tasks = (self.nb_classes - initial_increment) / increment
-        if not nb_tasks.is_integer():
-            raise Exception(
-                "The tasks won't have an equal number of classes"
-                f" with {len(self.class_order)} and increment {increment}"
-            )
-        increments.extend([increment for _ in range(int(nb_tasks))])
+
+            # Check if the total number of classes is compatible between increment list and self.nb_classes
+            if not (np.array(increment).sum()== self.nb_classes()):
```
Comment (arthurdouillard) [inline clloader/loader.py:59]: `if not sum(increment) == self.nb_classes` is simpler. The builtin `sum` does the trick, and because `self.nb_classes` has a property decorator, it doesn't need to be explicitly called.

File: clloader/scenario_loaders/classIncremental.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
+
```
Comment (arthurdouillard) [inline clloader/scenario_loaders/classIncremental.py:1]: I like the idea of having different kind of loaders for different scenario.

Altought I'd prefer the sub-package to be called `scenario` instead of `scenario_loaders`, shorter is better (*i.e.* PyTorch api > TF api :p)
And also rename the module in snake case instead of camel case.

Comment (arthurdouillard) [reply clloader/scenario_loaders/classIncremental.py:1]: And actually, the best would be to import the three scenario in the `__init__.py` of the `scenario` sub-package.

So the user would import likewise: `from clloader.scenario import ClassIncremental, InstanceIncremental`

I'm really fond of short import route ❤️

File: clloader/scenario_loaders/classIncremental.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    :param train_transformations: A list of data augmentation applied to the train set.
+    :param common_transformations: A list of transformations applied to both the
+                                   the train set and the test set. i.e. normalization,
+                                   resizing, etc.
+    :param evaluate_on: How to evaluate on val/test, either on all `seen` classes,
+                        on the `current` classes, or on `all` classes.
+    :param class_order: An optional custom class order, used for NC.
+    """
+
+    def __init__(
```
Comment (arthurdouillard) [inline clloader/scenario_loaders/classIncremental.py:33]: Do we want some kind of abstract scenarii defining common methods?

Then all subsequents scenario override only a few method, similar to what was done for datasets.

File: tests/Tests_Readme.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (arthurdouillard) [inline tests/Tests_Readme.md:1]: I'd rename this file simply `README.md` for consistency.

File: clloader/scenario_loaders/classIncremental.py (hunk 4 above)
Comment (arthurdouillard) [reply clloader/scenario_loaders/classIncremental.py:33]: Oh, was that your "classe CLLoader de base"? If so, nevermind.

File: clloader/datasets/continuumDataset.py (hunk 1 above)
Comment (TLESORT) [reply clloader/datasets/continuumDataset.py:11]: Oui j'ai compris ça après, je vais changer

File: clloader/scenario_loaders/classIncremental.py (hunk 4 above)
Comment (TLESORT) [reply clloader/scenario_loaders/classIncremental.py:33]: yes, this is the idea, CLLoader could be the scenarii abstract class

File: clloader/scenario_loaders/classIncremental.py (hunk 3 above)
Comment (TLESORT) [reply clloader/scenario_loaders/classIncremental.py:1]: snake case for module and camel case for classes?

Comment (arthurdouillard) [reply clloader/scenario_loaders/classIncremental.py]: I recall that what PEP08 advise.

File: clloader/loader.py [hunk 6]
```diff
@@ -3,76 +3,10 @@
...
-        return img
-
-    def __getitem__(self, index):
-        """Method used by PyTorch's DataLoaders to query a sample and its target."""
-        img = self.get_image(index)
-        y = self.y[index]
-        img = self.trsf(img)
-        return img, y
-
+from copy import copy
```
Comment (arthurdouillard) [inline clloader/loader.py:9]: The correct order of import is:

1. builtin packages (here `copy`)
newline
2. external packages (here `numpy` and `torch`)
newline
3. internal packages and module (here `clloader`)

You can set this automatically in most editors (mine is VSCode).

File: README.md [hunk 7]
```diff
@@ -18,21 +18,23 @@ And run!
 ```python
 from torch.utils.data import DataLoader
 
-from clloader import CLLoader
+from clloader import ClassIncremental, Permutation, Rotations
```
Comment (arthurdouillard) [inline README.md:21]: Don't import in the example stuff that you don't use.

File: clloader/loader.py [hunk 8]
```diff
@@ -101,6 +36,8 @@ def __init__(
         evaluate_on="seen",
         class_order=None
     ) -> None:
+        print("INIT!!!!!!!!!!!!!!!!!!!!!")
```
Comment (arthurdouillard) [inline clloader/loader.py:39]: Ahah, does it need to be kept? :p

File: README.md (hunk 7 above)
Comment (TLESORT) [reply README.md:21]: C'était pour montrer ce qui était possible d'import, mais oui ok

File: clloader/loader.py (hunk 8 above)
Comment (TLESORT) [reply clloader/loader.py:39]: hahaha, je l'ai viré mais j'ai du oublier de push

File: clloader/loader.py [hunk 9]
```diff
@@ -114,46 +51,88 @@ def __init__(
...
+        # compute task label
+        train_t = self._set_task_labels(train_y, self.increments)
+        test_t = self._set_task_labels(test_y, self.increments)
+
+        # Dataset with task label
+        self.train_data = (train_x, train_y, train_t)  # (data, class label, task label)
+        self.test_data = (test_x, test_y, test_t)  # (data, class label, task label)
+
+
+    def _set_task_labels(self, y: np.ndarray, increments: List[int])-> np.ndarray:
```
Comment (arthurdouillard) [inline clloader/loader.py:92]: It's more a `_get_task_labels`, you don't set anything here.

File: clloader/loader.py [hunk 10]
```diff
@@ -201,37 +180,37 @@ def __getitem__(self, task_index):
...
+            test = self._select_data_by_classes(0, self.nb_classes, split="test")
 
-        test_dataset = Dataset(*test, self.test_trsf, open_image=not self.cl_dataset.in_memory)
+        test_dataset = TaskSet(*test, self.test_trsf, open_image=not self.cl_dataset.in_memory)
 
         return train_dataset, test_dataset
 
-    def _select_data(self, min_class, max_class, split="train"):
+
+    def _select_data_by_task(self,ind_task: int, split: str="train"):
```
Comment (arthurdouillard) [inline clloader/loader.py]: Missing space after comma after self.

File: clloader/scenarios/instance_incremental.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    :param evaluate_on: How to evaluate on val/test, either on all `seen` classes,
+                        on the `current` classes, or on `all` classes.
+    :param class_order: An optional custom class order, used for NC.
+    """
+    # TODO
+
+
+    # Vanilla NI: data are given randomly, each task has all classes and each task has different instances
+    def _set_task_labels(self, y, increments):
+        return np.random.randint(self.nb_tasks, size=len(y))
```
Comment (arthurdouillard) [inline clloader/scenarios/instance_incremental.py:29]: How can you be sure that each task has all classes here?
It may not be true, or am I missing something?

Comment (arthurdouillard) [reply clloader/scenarios/instance_incremental.py:29]: Or maybe it's still a WIP to be handled in a next PR?

File: clloader/loader.py (hunk 9 above)
Comment (TLESORT) [reply clloader/loader.py]: I use this function to "create" the task label vector in the setup, that is why I put set. But I wasn't sure about the function name ^^

File: clloader/scenarios/instance_incremental.py (hunk 11 above)
Comment (TLESORT) [reply clloader/scenarios/instance_incremental.py:29]: I am not sure but, there is a very high possibility that this is the case ^^ p > 99,99 % for MNIST I think. We can add something to be sure

File: clloader/scenarios/class_incremental.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                      every task the amount of new classes.
+    :param initial_increment: A different task size applied only for the first task.
+                              Desactivated if `increment` is a list.
+    :param train_transformations: A list of data augmentation applied to the train set.
+    :param common_transformations: A list of transformations applied to both the
+                                   the train set and the test set. i.e. normalization,
+                                   resizing, etc.
+    :param class_order: An optional custom class order, used for NC.
+    """
+    """Continual Loader, generating datasets for the consecutive tasks.
```
Comment (arthurdouillard) [inline clloader/scenarios/class_incremental.py:26]: Duplicata of docs.

File: README.md (hunk 7 above)
Comment (arthurdouillard) [reply README.md:21]: I've fixed.

We could write more about those in the documentation.

File: clloader/scenarios/instance_incremental.py (hunk 11 above)
Comment (arthurdouillard) [reply clloader/scenarios/instance_incremental.py]: I've added an issue for this #17

File: clloader/scenarios/class_incremental.py (hunk 12 above)
Comment (arthurdouillard) [reply clloader/scenarios/class_incremental.py:25]: Fixed
